File: Largest Magic Square/Largest Magic Square.py
# Solution for Largest Magic Square in PY

class Solution:
    def largestMagicSquare(self, grid):

        rows = len(grid)
        cols = len(grid[0])

        # Row prefix sums
        row_sum = [[0] * cols for _ in range(rows)]
        for r in range(rows):
            row_sum[r][0] = grid[r][0]
            for c in range(1, cols):
                row_sum[r][c] = row_sum[r][c - 1] + grid[r][c]

        # Column prefix sums
        col_sum = [[0] * cols for _ in range(rows)]
        for c in range(cols):
            col_sum[0][c] = grid[0][c]
            for r in range(1, rows):
                col_sum[r][c] = col_sum[r - 1][c] + grid[r][c]

        # Try square sizes from largest to smallest
        for side in range(min(rows, cols), 1, -1):

            for i in range(rows - side + 1):
                for j in range(cols - side + 1):

                    # Expected sum from the first row
                    expected_sum = (
                        row_sum[i][j + side - 1] -
                        (row_sum[i][j - 1] if j > 0 else 0)
                    )

                    is_magic = True

                    # Check rows
                    for r in range(i + 1, i + side):
                        current_row_sum = (
                            row_sum[r][j + side - 1] -
                            (row_sum[r][j - 1] if j > 0 else 0)
                        )
                        if current_row_sum != expected_sum:
                            is_magic = False
                            break
                    if not is_magic:
                        continue

                    # Check columns
                    for c in range(j, j + side):
                        current_col_sum = (
                            col_sum[i + side - 1][c] -
                            (col_sum[i - 1][c] if i > 0 else 0)
                        )
                        if current_col_sum != expected_sum:
                            is_magic = False
                            break
                    if not is_magic:
                        continue

                    # Check diagonals
                    diag_sum = 0
                    anti_diag_sum = 0
                    for d in range(side):
                        diag_sum += grid[i + d][j + d]
                        anti_diag_sum += grid[i + d][j + side - 1 - d]

                    if diag_sum == expected_sum and anti_diag_sum == expected_sum:
                        return side

        return 1

File: Largest Magic Square/Largest Magic Square.java
// Solution for Largest Magic Square in JAVA

class Solution {
    public int largestMagicSquare(int[][] grid) {

        int rows = grid.length;
        int cols = grid[0].length;

        // Row prefix sums
        int[][] rowSum = new int[rows][cols];
        for (int r = 0; r < rows; r++) {
            rowSum[r][0] = grid[r][0];
            for (int c = 1; c < cols; c++) {
                rowSum[r][c] = rowSum[r][c - 1] + grid[r][c];
            }
        }

        // Column prefix sums
        int[][] colSum = new int[rows][cols];
        for (int c = 0; c < cols; c++) {
            colSum[0][c] = grid[0][c];
            for (int r = 1; r < rows; r++) {
                colSum[r][c] = colSum[r - 1][c] + grid[r][c];
            }
        }

        // Try square sizes from largest to smallest
        for (int side = Math.min(rows, cols); side >= 2; side--) {

            for (int i = 0; i + side - 1 < rows; i++) {
                for (int j = 0; j + side - 1 < cols; j++) {

                    // Expected sum from the first row
                    int expectedSum =
                        rowSum[i][j + side - 1] -
                        (j > 0 ? rowSum[i][j - 1] : 0);

                    boolean isMagic = true;

                    // Check all rows
                    for (int r = i + 1; r < i + side; r++) {
                        int currentRowSum =
                            rowSum[r][j + side - 1] -
                            (j > 0 ? rowSum[r][j - 1] : 0);

                        if (currentRowSum != expectedSum) {
                            isMagic = false;
                            break;
                        }
                    }
                    if (!isMagic) continue;

                    // Check all columns
                    for (int c = j; c < j + side; c++) {
                        int currentColSum =
                            colSum[i + side - 1][c] -
                            (i > 0 ? colSum[i - 1][c] : 0);

                        if (currentColSum != expectedSum) {
                            isMagic = false;
                            break;
                        }
                    }
                    if (!isMagic) continue;

                    // Check diagonals
                    int diagSum = 0;
                    int antiDiagSum = 0;

                    for (int d = 0; d < side; d++) {
                        diagSum     += grid[i + d][j + d];
                        antiDiagSum += grid[i + d][j + side - 1 - d];
                    }

                    if (diagSum == expectedSum &&
                        antiDiagSum == expectedSum) {
                        return side;
                    }
                }
            }
        }

        // 1x1 is always a magic square
        return 1;
    }
}
